src/OrbitGl/TimerTrack.cpp

```diff
@@ -101,42 +127,122 @@ void TimerTrack::UpdatePrimitives(uint64_t min_tick, uint64_t max_tick,
+          }
+        }
+
+        uint64_t function_id = timer_info.function_id();
 
-        bool is_visible_width = normalized_length * canvas->GetWidth() > 1;
         bool is_selected = &text_box == selected_textbox;
         bool is_highlighted = !is_selected &&
                               function_id != orbit_grpc_protos::kInvalidFunctionId &&
                               function_id == highlighted_function_id;
 
-        Vec2 pos(world_timer_x, world_timer_y);
-        Vec2 size(world_timer_width, GetTextBoxHeight(timer_info));
-        float z = GlCanvas::kZValueBox + z_offset;
-        const Color kHighlightColor(100, 181, 246, 255);
+        static const Color kHighlightColor(100, 181, 246, 255);
         Color color = is_highlighted ? kHighlightColor : GetTimerColor(timer_info, is_selected);
-        text_box.SetPos(pos);
-        text_box.SetSize(size);
 
-        auto user_data = std::make_unique<PickingUserData>(
-            &text_box, [&](PickingId id) { return this->GetBoxTooltip(id); });
+        bool is_visible_width = elapsed_us * inv_time_window * canvas->GetWidth() > 1;
 
         if (is_visible_width) {
-          if (!is_collapsed) {
-            SetTimesliceText(timer_info, elapsed_us, world_start_x, z_offset, &text_box);
-          }
-          batcher->AddShadedBox(pos, size, z, color, std::move(user_data));
+          WorldXInfo world_x_info_left_overlap =
+              ToWorldX(start_us, start_or_prev_end_us, inv_time_window, world_start_x, world_width);
+
+          WorldXInfo world_x_info_right_overlap =
+              ToWorldX(end_or_next_start_us, end_us, inv_time_window, world_start_x, world_width);
+
+          Vec3 top_left(world_x_info_left_overlap.world_x_start, world_timer_y + box_height, z);
+          Vec3 bottom_left(
+              world_x_info_left_overlap.world_x_start + world_x_info_left_overlap.world_x_width,
+              world_timer_y, z);
+          Vec3 top_right(world_x_info_right_overlap.world_x_start, world_timer_y + box_height, z);
+          Vec3 bottom_right(
+              world_x_info_right_overlap.world_x_start + world_x_info_right_overlap.world_x_width,
+              world_timer_y, z);
+
+          Triangle left_triangle(top_left, bottom_left, top_right);
+          auto user_data_left = std::make_unique<PickingUserData>(
+              &text_box, [&](PickingId id) { return this->GetBoxTooltip(id); });
+          batcher->AddShadedTriangle(left_triangle, color, std::move(user_data_left),
+                                     ShadingDirection::kLeftToRight);
+
+          Triangle triangle(bottom_left, top_right, bottom_right);
+          auto user_data_right = std::make_unique<PickingUserData>(
+              &text_box, [&](PickingId id) { return this->GetBoxTooltip(id); });
+          batcher->AddShadedTriangle(triangle, color, std::move(user_data_right),
+                                     ShadingDirection::kRightToLeft);
+
```

Comment:
The overlap code is taking a significant portion of the function. Overlaps are only allowed for a very specific type of track, yet its logic is run for all tracks. Could we have an UpdatePrimitives method specifically for tracks that allow overlap and keep the original implementation for all other tracks?